File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
import sys
input = sys.stdin.readline


def solve(n: int, lst: list):
    stack = []
    rtn = 0 

    for i, h in enumerate(lst):
        while stack and stack[-1][1] > h:
            j, v = stack.pop()
            width = i if not stack else i - stack[-1][0] - 1
            rtn = max(rtn, width*v)

        stack.append([i,h])

    while stack:
        j, v = stack.pop()
        width = n if not stack else n - stack[-1][0] - 1
        rtn = max(rtn, width*v)
    
    return rtn

if __name__ == '__main__':
    Arr = list(map(int,input().split()))
    while Arr[0]:
        print(solve(Arr[0],Arr[1:]))
        Arr = list(map(int,input().split()))

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.js
function histogram(n, lst){
    let rtn = 0;
    const stack = [];

    lst.forEach((h, i) => {
        while (stack.length !== 0 && stack[stack.length-1][1] > h ){
            let [idx, height] = stack.pop();
            let width = stack.length !== 0 ? i - stack[stack.length-1][0] - 1 : i;
            rtn = Math.max(rtn, width * height);
        }
        stack.push([i, h]);
    });

    while (stack.length !== 0){
        let [idx, height] = stack.pop();
        let width = stack.length !== 0 ? n - stack[stack.length-1][0] - 1 : n;
        rtn = Math.max(rtn, width * height);
    }
    return rtn;
}

function solve(input) {
  // TODO: 알고리즘 로직
    for(let i=0; i < input.length -1; i++){
        const Arr = input[i].split(' ').map(Number);
        console.log(histogram(Arr[0], Arr.slice(1)));
    }
}


// 제출하기전에 false 로
const isLocal = false;  // 여기서 true로 설정하면 로컬 환경으로 간주

if (isLocal) {
    // 로컬 환경 (VSCode)에서의 입력 처리
    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    let lines = [];
    rl.on('line', (line) => {
        lines.push(line);
    }).on('close', () => {
        solve(lines);
    });
} else {
    // 백준 환경에서의 입력 처리
    const fs = require('fs');
    const lines = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
    solve(lines);
}